xds/src/main/java/io/grpc/xds/Filter.java

```diff
@@ -0,0 +1,136 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * Defines the parsing functionality of an HTTP filter. A Filter may optionally implement either
+ * {@link ClientInterceptorBuilder} or {@link ServerInterceptorBuilder} or both, indicating it is
+ * capable of working on the client side or server side or both, respectively.
+ */
+interface Filter {
+
+  /**
+   * The proto message types supported by this filter. A filter will be registered by each of its
+   * supported message types.
+   */
+  String[] typeUrls();
+
+  /**
+   * Parses the top-level filter config from raw proto message. The message may be either a {@link
+   * com.google.protobuf.Any} or a {@link com.google.protobuf.Struct}.
+   */
+  StructOrError<? extends FilterConfig> parseFilterConfig(Message rawProtoMessage);
+
+  /**
+   * Parses the per-filter override filter config from raw proto message. The message may be either
+   * a {@link com.google.protobuf.Any} or a {@link com.google.protobuf.Struct}.
+   */
+  StructOrError<? extends FilterConfig> parseFilterConfigOverride(Message rawProtoMessage);
+
+  /** Represents an opaque data structure holding configuration for a filter. */
+  interface FilterConfig {
+    String typeUrl();
+  }
+
+  /** Uses the FilterConfigs produced above to produce an HTTP filter interceptor for clients. */
+  interface ClientInterceptorBuilder {
+    @Nullable
+    ClientInterceptor buildClientInterceptor(
+        FilterConfig config, @Nullable FilterConfig overrideConfig, PickSubchannelArgs args,
+        ScheduledExecutorService scheduler);
+  }
+
+  // Server side filters are not currently supported, but this interface is defined for clarity.
+  interface ServerInterceptorBuilder {
+    ServerInterceptor buildServerInterceptor(
+        FilterConfig config, @Nullable FilterConfig overrideConfig);
+  }
+
+  final class StructOrError<T> {
+    /**
+     * Returns a {@link StructOrError} for the successfully converted data object.
+     */
+    static <T> StructOrError<T> fromStruct(T struct) {
+      return new StructOrError<>(struct);
+    }
+
+    /**
+     * Returns a {@link StructOrError} for the failure to convert the data object.
+     */
+    static <T> StructOrError<T> fromError(String errorDetail) {
+      return new StructOrError<>(errorDetail);
+    }
+
+    final String errorDetail;
+    final T struct;
+
+    private StructOrError(T struct) {
+      this.struct = checkNotNull(struct, "struct");
+      this.errorDetail = null;
+    }
+
+    private StructOrError(String errorDetail) {
+      this.struct = null;
+      this.errorDetail = checkNotNull(errorDetail, "errorDetail");
+    }
+  }
+
+  /**
+   * A registry for all supported {@link Filter}s. Filters can be queried from the registry
+   * by any of the {@link Filter#typeUrls(), type URLs}.
+   */
+  final class Registry {
+    static Registry GLOBAL_REGISTRY =
```

Comment:
nit: `GLOBAL_REGISTRY`. I am not a fan of such a name, it looks strange to me... `FilterRegistry.getDefaultRegistry()` looks much more comfortable to me...

Reply:
Done